File: 2724-convert-an-array-into-a-2d-array-with-conditions/convert-an-array-into-a-2d-array-with-conditions.cpp
class Solution {
public:
    vector<vector<int>> findMatrix(vector<int>& nums) {
        unordered_map<int,int> freq;
        int maxi = 0;
        for(int num: nums) // o(n) always
        {
            freq[num]++;
            maxi = max(maxi, freq[num]);
        }
        vector<vector<int>> result(maxi);
        for(auto [key, value]: freq) // o(n) worst case
            for(int ind=0; ind<value; ind++)
                result[ind].push_back(key);
        return result;
    }
};

File: take-u-forward/2724-convert-an-array-into-a-2d-array-with-conditions/convert-an-array-into-a-2d-array-with-conditions.java
class Solution {
    public List<List<Integer>> findMatrix(int[] nums) {
        List<List<Integer>> res = new ArrayList();
        int freq[] = new int[nums.length + 1];
        int max = 0;
        for(int num: nums)
        {
            if(++freq[num] > max)
            {
                max++;
                res.add(new ArrayList<Integer>());
            }
            res.get(freq[num]-1).add(num);
        }
        return res;
    }
}